e2etest/bgptests/bgp.go

```diff
@@ -147,6 +147,115 @@ var _ = ginkgo.Describe("BGP", func() {
 			}),
 	)
 
+	ginkgo.Describe("GracefulRestart, when speakers restart", func() {
+
+		var grEnabled bool
+
+		ginkgo.AfterEach(func() {
+			for _, c := range FRRContainers {
+				c.NeighborConfig.GracefulRestart = false
+			}
+		})
+
+		assertDuringSpeakerRestart := func(pairingIPFamily ipfamily.Family, poolAddresses []string, tweak testservice.Tweak) {
+			_, svc := setupBGPService(cs, testNamespace, pairingIPFamily, poolAddresses,
+				FRRContainers, func(svc *corev1.Service) {
+					testservice.TrafficPolicyCluster(svc)
+					tweak(svc)
+				})
+			defer testservice.Delete(cs, svc)
+
+			allNodes, err := cs.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			testservice.ValidateDesiredLB(svc)
+
+			for _, c := range FRRContainers {
+				validateService(svc, allNodes.Items, c)
+			}
+			err = metallb.RestartSpeakerPodsNoWait(cs)
+			Expect(err).NotTo(HaveOccurred())
+
+			if !grEnabled {
+				Eventually(func() error {
+					for _, c := range FRRContainers {
+						err := validateServiceNoWait(svc, allNodes.Items, c)
+						if err != nil {
+							if errors.Is(err, ErrStaleRoute) {
+								Expect(err).NotTo(HaveOccurred(),
+									"a stale route cannot be observed if GR disabled")
+							}
+							return err
+						}
+					}
+					return nil
+				}, 2*time.Minute, time.Second).Should(HaveOccurred(), "a downtime should be observed")
+				return
+			}
+
+			Eventually(func() error {
+				for _, c := range FRRContainers {
+					err := validateServiceNoWait(svc, allNodes.Items, c)
+					if err != nil {
+						if errors.Is(err, ErrStaleRoute) {
+							continue // when GR, is normal to observe stale routes
+						}
+						Expect(err).NotTo(HaveOccurred(), "downtime was observed")
+					}
+				}
+
+				pods, err := metallb.SpeakerPods(cs)
+				if err != nil {
+					return err
+				}
+
+				for _, p := range pods {
+					if !k8s.PodIsReady(p) {
+						return fmt.Errorf("speaker pods are not ready")
+					}
+				}
+
+				return nil
+			}, 2*time.Minute, time.Second).ShouldNot(HaveOccurred(), "no downtime until speakers are ready")
+
+			for _, c := range FRRContainers {
+				validateService(svc, allNodes.Items, c)
+			}
+		}
+
+		ginkgo.Context("and when GR enabled", func() {
+			ginkgo.BeforeEach(func() {
+				grEnabled = true
```

Comment:
let's have this a member of the assertDuringSpeakerRestart function (shouldFail, for example)

Reply:
not done, can you expand what you mean?

Reply:
instead of having a global that share some state, I'd make assertDuringSpeakerRestart pure with that regard and allow it to accept an extra parameter which is that bool. The same bool can then be a parameter of the table tests (bonus point, use a constant with the proper name instead of a bool).

Reply:
Currently I have two tables, exactly the same with the diff the beforeeach section. I think the way that the Describe scope have a "global" state, and that in the before each you define it (as you do with the frr config. gr to true) is more readable to understand the test behavior and fits better ginkgo approach. The description looks like 

```
BGP GracefulRestart, when speakers restart and when GR enabled dataplane should keep working (entry x,y,z)
BGP GracefulRestart, when speakers restart and when GR disabled dataplane should have a downtime (entry x,y,z)
```

If I add as extra to the function, then you suggest to actually merge the test into single context/table (right?) and have the  bool name hint  what we test inside the entry  args instead of the ginkgo description. tbh I would prefer to keep it that way but I can change it if you want.